contrib/kube-prometheus/manifests/prometheus/prometheus-k8s-rules.yaml

```diff
@@ -516,4 +516,52 @@ data:
         summary = "Kubelet is close to pod limit",
         description = "Kubelet {{$labels.instance}} is running {{$value}} pods, close to the limit of 110",
       }
-    
+
+    ALERT K8SPodWaiting
+      IF sum(sum_over_time(kube_pod_container_status_waiting[5m]) > 0 ) BY (job, exported_pod, instance, namespace)
+      FOR 5m
+      LABELS {
+        service = "k8s",
+        severity = "critical",
+      }
+      ANNOTATIONS {
+        summary = "Pod is waiting too long",
+        description = "Pod {{ $labels.namespace }}/{{ $labels.exported_pod }} is in waiting state",
+      }
+
+    ALERT K8sPodRestartingTooMuch
+      IF sum (rate(kube_pod_container_status_restarts[5m]) > 0) BY(job, exported_pod, instance, namespace)
+      FOR 5m
```

Comment:
Yes, both alerts are on the pod level and could get pretty noisy. Critical/pages should be dedicated to service level objectives, i.e. "service is responding slowly". Pods not starting etc. are potential causes and should be a warning at max.

I'm wondering here how accurate the expression is in general. Rate will only return to 0 after the last restart is 5 minutes in the past, meaning the alert will resolve rather late.  
But it might also stop firing because the restart backoff got quite long, which can cause the alert to flap on and off between restart attempts.

Does it maybe make sense to catch all non-ready pods more generally? I think ultimately we want to know what ports are not running and the exact failure mode is something we'd look up with  `kubectl` anyway.